s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -58,6 +60,32 @@ private[alpakka] final class S3Stream(credentials: AWSCredentials, region: Strin
   val MinChunkSize = 5242880 //in bytes
   val signingKey = SigningKey(credentials, CredentialScope(LocalDate.now(), region, "s3"))
 
+  /** listBucket lists files in a bucket, filtered by an optional prefix 
+    * 
+    * @param bucket name of the bucket
+    * @param prefix optional prefix
+    * @return A akka Source with keys
+    * 
+    */
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[String, NotUsed] = {
```

Comment:
Nice! Always fun to compare implementations. We are also using a fork of alpakka internally to get listBucket functionality.

1) I use prefix to limit the number of keys returned. But use .take(k) if you want to cancel stream after k keys.
2) Yes, first page emits all its elements before getting next page because of the use of concat (http://doc.akka.io/api/akka/2.4/akka/stream/scaladsl/Source.html#concat)